Source/ExecutionEngine/ExecutionEngine.cs

```diff
@@ -706,21 +706,17 @@ private Implementation[] GetPrioritizedImplementations(Program program)
       }
 
       await verifyImplementationSemaphore.WaitAsync(cts.Token);
-      try {
-
-        ImplIdToCancellationTokenSource.AddOrUpdate(id, cts, (k, ov) => cts);
+      ImplIdToCancellationTokenSource.AddOrUpdate(id, cts, (k, ov) => cts);
 
-        var coreTask = Task.Run(() => VerifyImplementation(processedProgram, stats, er, cts.Token,
-          implementation,
-          programId, taskWriter), cts.Token);
+      var coreTask = Task.Run(() => VerifyImplementation(processedProgram, stats, er, cts.Token,
+        implementation,
+        programId, taskWriter), cts.Token);
 
-        var verificationResult = await coreTask;
-        return verificationResult;
-      }
-      finally {
+      var _ = coreTask.ContinueWith(t => {
         verifyImplementationSemaphore.Release();
         ImplIdToCancellationTokenSource.TryRemove(id, out old);
```

Comment:
Which thread does `ContinueWith` run on?  Is it safe to release on the thread?

Reply:
Where the continuation runs is scheduled by `TaskScheduler.Current`. Does it matter?

> Is it safe to release on the thread?

Why wouldn't it be? `verifyImplementationSemaphore.Release();` is thread-safe.